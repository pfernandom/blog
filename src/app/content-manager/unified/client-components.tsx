'use client'

import Image, { ImageLoader } from 'next/image'
import React from 'react'

const imageLoader: ImageLoader = ({ src, width, quality }) => {
  return `/${src}?w=${width}&q=${quality || 75}`
}

export default function Img({
  src,
  alt,
  width,
  height,
}: {
  src: string | undefined
  alt: string | undefined
  width: `${number}` | number | undefined
  height: `${number}` | number | undefined
}) {
  if (!src || !alt) {
    // eslint-disable-next-line @next/next/no-img-element
    return <img src={src} alt={alt} />
  }
  return (
    <Image
      loader={imageLoader}
      src={src}
      alt={alt}
      width={width}
      height={height}
    />
  )
}
