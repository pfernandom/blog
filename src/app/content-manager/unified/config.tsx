import Img from './client-components'
import {
  ButtonHTMLAttributes,
  DetailedHTMLProps,
  ImgHTMLAttributes,
} from 'react'
import { type Options } from 'rehype-react'
import { createElement, Fragment } from 'react'
import path from 'path'
import type { Element } from 'hast'
import React from 'react'

type WithNode = {
  node: Element
}

type ImageProperties =
  | { alt: string; width: `${number}` | number; height: `${number}` | number }
  | undefined

export const MyImage = ({
  src,
  alt,
  width,
  height,
}: DetailedHTMLProps<ImgHTMLAttributes<HTMLImageElement>, HTMLImageElement> &
  ImageProperties) => {
  const rightSrc = path.join('content', src ?? 'noimage.jpg')
  return <Img src={rightSrc} alt={alt} width={width} height={height} />
}

export function buildOptions(): Options {
  return {
    createElement,
    Fragment,
    components: {
      img: ({
        src,
        alt,
        node,
      }: DetailedHTMLProps<
        ImgHTMLAttributes<HTMLImageElement>,
        HTMLImageElement
      > &
        WithNode) => {
        const properties = node.properties as ImageProperties
        return (
          <Img
            src={src}
            alt={properties?.alt ?? alt}
            width={properties?.width}
            height={properties?.height}
          />
        )
      },
      p: ({ children }) => <p>{children}</p>,
      // table: (args) => <div>TABLE</div>,
      button: ({
        children,
        onClick,
      }: DetailedHTMLProps<
        ButtonHTMLAttributes<HTMLButtonElement>,
        HTMLButtonElement
      >) => <button onClick={onClick}>{children}</button>,
    },
    passNode: true,
  }
}
