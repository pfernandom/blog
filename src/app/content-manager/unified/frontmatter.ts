import { Post } from '../manager'
import { format, parseISO } from 'date-fns'
import UrlManager from '../url-manager'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function validatePostData(data: { [key: string]: any }): Post {
  const {
    slug,
    title,
    date,
    description,
    key_words,
    hero_image,
    hero_image_original,
    hero_image_blur,
    hero_image_alt,
    hero_width,
    hero_height,
    published,
    series,
    social_title,
    social_subtitle,
    social_footer,
    test,
    legacy,
    subblog_slug,
  } = data

  const postData = {
    slug,
    title,
    date: format(parseISO(date), 'MMMM dd, yyyy'),
    description,
    key_words,
    hero_image,
    hero_image_original,
    hero_image_blur,
    hero_image_alt,
    hero_width,
    hero_height,
    published,
    series,
    social_title,
    social_subtitle,
    social_footer,
    test,
    legacy,
    subblog_slug,
  }

  const urlManager = new UrlManager()
  const fullSlug = urlManager.getSlugFromPostInfo(postData)
  postData.slug = fullSlug

  return postData
}
