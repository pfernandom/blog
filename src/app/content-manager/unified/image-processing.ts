import fs from 'node:fs'
import path from 'path'
import sharp from 'sharp'

/**
 * Parse attributes like:
 * ![alt="Hero image",width=100,height=100](./hero.jpeg)
 * @param src the image source
 * @returns a map of properties or undefined
 */
export function parseImageSource(
  src: string
): Record<string, string> | undefined {
  return src
    .match(/(\w+?=[\w\s\d"]+)*/g)
    ?.filter((el) => el.length)
    .map((el) => el.split('='))
    .reduce((acc, el) => ({ [el[0]]: el[1], ...acc }), {})
}

/**
 * Move image to public folder, and perform optimizations on it
 * @param filePath the original image path
 * @param outputPath the output path for the image
 * @param width optional width for resizing
 * @param height optional height for resizing
 */
function resizeAndMoveToPublic(
  filePath: string,
  outputPath: string,
  width?: number | null,
  height?: number | null
) {
  try {
    const sh = sharp(filePath)
    if (width && height) {
      sh.resize({
        width,
        height,
        fit: 'cover',
      })
    }
    sh.toFile(outputPath)
  } catch (error) {
    throw new Error(`Could not process image ${filePath}: \n${error}`)
  }
}

/**
 * Extract image properties from the Markdown "alt" text field.
 *
 * @param alt the alt text field, possibly with properties
 * @returns a map of properties, if they exist
 */
function getMaybeImageProperties(alt?: string | null) {
  const maybeProperties = parseImageSource(alt as string)
  if (!maybeProperties) {
    return null
  }

  return {
    width: maybeProperties['width'] ? parseInt(maybeProperties['width']) : null,
    height: maybeProperties['height']
      ? parseInt(maybeProperties['height'])
      : null,
    alt: maybeProperties['alt'],
  }
}

interface MDImageProperties {
  width: number | null
  height: number | null
  alt: string
}

export function moveAndProcessImage(
  contentPath: string,
  src: string,
  alt: string | undefined | null,
  fileNameMap: Record<string, string>
): [MDImageProperties | null, string] {
  const filePath = path.join(path.dirname(contentPath), src as string)
  const publicPath = path.join('static', filePath)
  const outputPath = path.join('public', publicPath)

  fs.mkdirSync(path.dirname(outputPath), { recursive: true })
  if (
    fileNameMap[`${contentPath}_${src}`] &&
    fileNameMap[`${contentPath}_${src}`] !== filePath
  ) {
    fs.unlinkSync(fileNameMap[`${contentPath}_${src}`])
  }
  fileNameMap[`${contentPath}_${src}`] = filePath

  const maybeProperties = getMaybeImageProperties(alt)

  if (maybeProperties && maybeProperties.width && maybeProperties.height)
    resizeAndMoveToPublic(
      filePath,
      outputPath,
      maybeProperties.width,
      maybeProperties.height
    )

  // node.properties.src = publicPath
  return [maybeProperties, publicPath]
}
