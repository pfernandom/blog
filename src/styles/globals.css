html,
body {
  padding: 0;
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen,
    Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;
  /* font: 100%/1.75 "Avenir Next", "Helvetica Neue", "Segoe UI", "Helvetica",
    "Arial", sans-serif; */
}

footer {
  text-align: center;
}

code {
  background: #f5f2f0;
  padding: 0em 0.5em;
  display: inline-block;
  font-family: Roboto;
  font-size: 1.1em;
}

@media (prefers-color-scheme: dark) {
  html {
    color-scheme: dark;
  }
  body {
    color: white;
    background: black;
  }
  code {
    background-color: grey;
  }
  pre {
    filter: brightness(0.95);
  }
}

h1,
h2,
h3,
h4 {
  font-family: Montserrat, sans-serif;
  font-weight: 500;
  text-rendering: optimizeLegibility;
  line-height: 1.1;
}

h2,
h3,
h4 {
  margin-top: 2em;
}

h1 {
  font-size: 2em;
  color: var(---heading-color);
}

h2 {
  font-size: 1.5em;
}

h2,
h3 {
  color: var(---heading-color);
  filter: brightness(0.8);
}

a:hover {
  text-decoration: underline;
}

p {
  margin-bottom: 1.5em;
}

.layout-content {
  max-width: 900px;
  margin: 1em auto;
  padding: 0 2em;
}

a {
  color: var(---heading-color);
}

a:hover {
  transform: scale(1.05);
  display: inline-block;
}

@import url(./blog.css);
@import url(./instagram.css);
