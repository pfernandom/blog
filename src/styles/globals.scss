html,
body {
  padding: 0;
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen,
    Ubuntu, Cantarell, 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;

  /* font: 100%/1.75 "Avenir Next", "Helvetica Neue", "Segoe UI", "Helvetica",
    "Arial", sans-serif; */
}

footer {
  text-align: center;
}
/* 
code {
  background: #f5f2f0;
  padding: 0 0.5em;
  display: inline-block;
  font-family: Roboto, sans-serif;
  font-size: 1.1em;
} */

:not(pre) code {
  margin: 1px;
  padding-left: 1px;
  padding-right: 1px;
}

@media (prefers-color-scheme: dark) {
  html {
    color-scheme: dark;
  }

  body {
    color: white;
    background: black;
  }

  code {
    background-color: grey;
  }

  pre {
    filter: brightness(0.95);
  }
}

h1,
h2,
h3,
h4 {
  font-family: Montserrat, sans-serif;
  font-weight: 500;
  text-rendering: optimizelegibility;
  line-height: 1.1;
}

h2,
h3,
h4 {
  margin-top: 2em;
}

h1 {
  font-size: 2em;
  color: var(---heading-color);
}

h2 {
  font-size: 1.5em;
}

h2,
h3,
h4 {
  color: var(---heading-color);
  filter: brightness(0.8);
}

a:hover {
  text-decoration: underline;
}

p {
  margin-bottom: 1.5em;
}

.layout-content {
  max-width: 900px;
  margin: 1em auto;
  padding: 0 2em;
}

a {
  color: var(---heading-color);
  text-decoration: none;
}

a:hover {
  text-decoration: underline;
}

@import './blog.scss';
@import './instagram.scss';

blockquote {
  --background: #fff;
  --color: #00cc8f;
  --font-size: 1.2em;

  /* padding: 0.5em;
  background-color: gainsboro;
  font-family: Courier; */
  border-left: 3px solid var(--color);
  color: #1a1a1a;
  font-family: Georgia, Times, 'Times New Roman', serif;
  font-size: var(--font-size);
  font-style: italic;
  line-height: 1.8em;
  margin-block-start: 0;
  margin-block-end: 0;
  margin-left: 0;
  padding: 1em 2em;
  position: relative;
  transition: 0.2s border ease-in-out;
  z-index: 0;
  width: 80%;
}

blockquote p {
  margin: 0;
  margin-bottom: 0.5em;
}

blockquote::before {
  content: '';
  position: absolute;
  top: 50%;
  left: -4px;
  height: 2em;
  background-color: var(--background);
  width: 5px;
  margin-top: -1em;
}

blockquote::after {
  content: '\201C';
  position: absolute;
  top: 50%;
  left: -0.5em;
  color: var(--color);
  font-style: normal;
  font-size: 2em;
  font-weight: bold;
  line-height: 1.5em;
  text-align: center;
  text-indent: -2px;
  width: 1em;
  margin-top: -0.5em;
  transition: 0.2s all ease-in-out, 0.4s transform ease-in-out;
}

blockquote.active:after {
  transform: rotateY(360deg);
}

blockquote h3 {
  margin-top: 0;
  font-size: var(--font-size);
}
