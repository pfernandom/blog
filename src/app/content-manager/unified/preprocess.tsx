import { Root } from 'hast'
import fs from 'node:fs'
import { Transformer } from 'unified'
import { visit } from 'unist-util-visit'
import { moveAndProcessImage } from './image-processing'

const fileMapPath = '.next/content_map.json'

export function preprocessPlugin({
  contentPath,
}: {
  contentPath: string
}): void | Transformer<Root, Root> {
  return (tree) => {
    const fileNameMap: Record<string, string> = fs.existsSync(fileMapPath)
      ? JSON.parse(fs.readFileSync(fileMapPath, 'utf8'))
      : {}
    visit(tree, 'element', (node) => {
      switch (node.tagName) {
        case 'img':
          const src = node.properties?.src
          const alt = node.properties?.alt

          if (node.properties && src && typeof src === 'string') {
            const [maybeImageProperties, publicPath] = moveAndProcessImage(
              contentPath,
              src,
              alt as string | null | undefined,
              fileNameMap
            )
            node.properties = { ...maybeImageProperties, ...node.properties }
            node.properties.src = publicPath
          }
      }
    })

    fs.writeFileSync(fileMapPath, JSON.stringify(fileNameMap, null, 2))
  }
}
