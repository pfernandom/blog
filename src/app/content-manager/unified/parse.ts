import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import { unified } from 'unified'
// import { reactPlugin, reactOutputPlugin } from './unified'
import rehypeReact from 'rehype-react'
import { buildOptions } from './config'
import { preprocessPlugin } from './preprocess'
// import type {IntrinsicElements} from 'react'

import rehypeInlineCodeClassNamePlugin from 'rehype-inline-code-classname'
import rePrism from './re-prism'
import remarkGfm from 'remark-gfm'

export function parseMarkdown(path: string, content: string) {
  return unified()
    .use(remarkParse)
    .use(remarkRehype, { allowDangerousHtml: true })
    .use(remarkGfm)
    .use(rehypeInlineCodeClassNamePlugin)
    .use(rePrism)
    .use(preprocessPlugin, { contentPath: path })
    .use(rehypeReact, buildOptions())
    .processSync(content).result
}
