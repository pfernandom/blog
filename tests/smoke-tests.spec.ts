import { describe, expect, test } from '@jest/globals'
import { fromFrontMatter } from 'app/content-manager/unified/frontmatter'
import { globSync } from 'glob'
import fs from 'node:fs'
import path from 'path'

describe('Smoke tests', function () {
  test('Snapshot: All posts have the expected information', async function () {
    const cutoutDate = new Date('2023-07-22')

    const staticPosts = globSync(path.join('content/**/*.md'))
    const mdxPosts = globSync(path.join('src', 'blog', 'content/**/*.mdx'))

    const posts = [...staticPosts, ...mdxPosts]
      .map((p) => fs.readFileSync(p, 'utf8'))
      .map((p) => fromFrontMatter(p))
      .filter(
        (content) =>
          (content?.date ? new Date(content.date) : new Date()).getTime() <
          cutoutDate.getTime()
      )

    expect(posts).toMatchSnapshot()
  })
})
