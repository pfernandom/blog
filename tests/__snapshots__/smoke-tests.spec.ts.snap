// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Smoke tests Snapshot: All posts have the expected information 1`] = `
[
  {
    "date": "1988-06-19",
    "description": [
      "line 1",
      "line 2",
    ],
    "hero_image": "./hero.jpeg",
    "hero_image_alt": "A hero image for this post about state management",
    "key_words": "test, post",
    "legacy": true,
    "published": true,
    "slug": "blog/1988/6/mypost",
    "test": true,
    "title": "This is a test post",
  },
]
`;
